### PR TITLE
add copy on write

### PR DESCRIPTION
**Pull request by kpango**
Signed-off-by: kpango <kpango@vdaas.org>

### Description:
Copy on Write functionality has been implemented.

Vald Agent's Copy on Write performs backups using three directories.
1. tempdir
2. index path
3. old index path
In this case, the backup data of the old Index 3 is overwritten only once, without generation management this time.

The actual operation is as follows only when CoW is enabled.
1. Vald Agent writes the indexes to tempdir at the time of SaveIndex.
2. Secondly, Vald Agent moves the data in the index path, which is the current backup data, to the old index path when the exporting is completed.
3. Finally, after moving the index written to tempdir to the main databank (index path), a new tempdir is created and the export is complete.

### Environment:

- Go Version: 1.17.7
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 1.14.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by vankichi — APPROVED** [review]
LGTM...I'll wait for success operation verification.

File: internal/info/info_test.go [hunk 1]
```diff
@@ -601,7 +601,7 @@ func Test_info_String(t *testing.T) {
 				},
 			},
 			want: want{
-				want: "\nbuild cpu info flags ->\t[]\nbuild time           ->\tbt\ncgo enabled          ->\ttrue\ngit commit           ->\tcommit\ngo arch              ->\tgoarch\ngo os                ->\tgoos\ngo version           ->\t1.1\nngt version          ->\t1.2\nserver name          ->\tsrv\nstack trace-0        ->\turl\tfunc\nvald version         ->\t\x1b[1m1.0\x1b[22m",
+				want: "\nbuild cpu info flags ->\t[]\nbuild time           ->\tbt\ncgo enabled          ->\ttrue\ngit commit           ->\tcommit\ngo arch              ->\tgoarch\ngo os                ->\tgoos\ngo version           ->\t1.1\nngt version          ->\t1.2\nserver name          ->\tsrv\nstack trace-0        ->\turl\tfile#L10\tfunc\nvald version         ->\t\x1b[1m1.0\x1b[22m",
```
Comment (kpango) [inline internal/info/info_test.go:604]:
```suggestion
				want: "\nbuild cpu info flags ->\t[]\nbuild time           ->\tbt\ncgo enabled          ->\ttrue\ngit commit           ->\tcommit\ngo arch              ->\tgoarch\ngo os                ->\tgoos\ngo version           ->\t1.1\nngt version          ->\t1.2\nserver name          ->\tsrv\nstack trace-000        ->\turl\tfile#L10\tfunc\nvald version         ->\t\x1b[1m1.0\x1b[22m",
```

File: internal/info/info_test.go [hunk 2]
```diff
@@ -713,7 +713,7 @@ func TestDetail_String(t *testing.T) {
...
```
Comment (kpango) [inline internal/info/info_test.go:716]:
```suggestion
				want: "\nbuild cpu info flags ->\t[]\nbuild time           ->\tbt\ncgo enabled          ->\ttrue\ngit commit           ->\tcommit\ngo arch              ->\tgoarch\ngo os                ->\tgoos\ngo version           ->\t1.1\nngt version          ->\t1.2\nserver name          ->\tsrv\nstack trace-000        ->\turl\tfile#L10\tfunc\nvald version         ->\t\x1b[1m1.0\x1b[22m",
```

File: pkg/agent/core/ngt/service/ngt.go [hunk 3]
```diff
@@ -93,7 +92,7 @@ type ngt struct {
 	indexing  atomic.Value
 	saving    atomic.Value
 	cimu      sync.Mutex // create index mutex
-	lastNoice uint64     // last number of create index execution this value prevent unnecessary saveindex.
+	lastNocie uint64     // last number of create index execution this value prevent unnecessary saveindex.
```
Comment (vankichi) [inline pkg/agent/core/ngt/service/ngt.go:95]: Does it suppose `lastNotice`? :thinking:

Comment (kpango) [reply pkg/agent/core/ngt/service/ngt.go:95]: lastNotice is typo.

as you can see in the same struct field  nocie.
NOCIE means  (Number Of Create Index Execution)

so lastNocie also same

lastNocie  = Last Number Of Create Index Execution

Comment (vankichi) [reply pkg/agent/core/ngt/service/ngt.go:95]: Got it !

**Review by kevindiu — APPROVED** [review]

**Review by vankichi — APPROVED** [review]
LGTM :+1:

**Review by hlts2 — APPROVED** [review]
LGTM 👍

**Pull request merged**